### ISSUE TITLE
Interactive Progress in Console Not Working

### ISSUE THREAD

**Issue by MattParkerDev** [thomhurst/ModularPipelines#452]
Hi,

I'm using Modular Pipelines on windows (Windows Terminal with Powershell 7)~~, and inside Github Runners (Ubuntu Latest).~~

I cannot get the interactive console progress to work - the console just seems to hang while it does work in the background, and then displays all of the results once complete. Am I missing something?

![image](https://github.com/thomhurst/ModularPipelines/assets/61717342/f5f3ca62-6c51-4c63-80ff-b6be3718f42a)
Figure: Hangs at this point

I noticed all of the logs which are displayed once it unhangs, all have the same timestamp.

**Comment by thomhurst** [comment]
Hi @MattParkerDev!

GitHub runners don't  support the 'interactive console' mode so you won't see any progress bars for this. However Windows and Powershell should support it. When in interactive mode you should see progress bars.

For non interactive mode, ModularPipelines does do lazy logging, which means that you won't see logs for a module until that module has finished. It's designed this way, so that logs can be grouped together. Otherwise, with everything running in parallel, it'd be hard to tell what log related to which module. So logs for a particular module will all have the same timestamp. (I might have to look into improving this)

However, if ALL logs for all modules aren't happening until the very end, that does seem like something strange is happening.

Ideally I'd need a small reproduction app to help diagnose any issues.

**Comment by MattParkerDev** [comment]
When you say interactive mode, I'm not sure how to enable that in windows/powershell :)

**Comment by MattParkerDev** [comment]
Yes, all of the logs for all of the modules are not displayed, until all of the modules are complete, and then they are all displayed together.

**Comment by thomhurst** [comment]
Thanks very much! I've set up a discussions section in GitHub if you're happy to post there? 
https://github.com/thomhurst/ModularPipelines/discussions

**Comment by thomhurst** [comment]
As for interactive mode, it should be detected and enabled automatically if supported.

Can you try setting this on startup and seeing if it improves anything?

`AnsiConsole.Profile.Capabilities.Interactive = false;`

Just to eliminate whether interactive logging isn't working properly.

**Comment by MattParkerDev** [comment]
This code seems to be the culprit in my case. Disposing of the module writes its logs to the console.
However with ShowProgressInConsole set to true, the module is not disposed once it completes. Is there a different mechanism for outputting the logs in interactive mode that doesn't seem to be working here?

Setting ShowProgressInConsole to `false` causes `await _moduleDisposer.DisposeAsync(module);` to be called, and I see logs for modules individually as they finish.

```csharp
finally
        {
            if (!AnsiConsole.Profile.Capabilities.Interactive || !_pipelineOptions.Value.ShowProgressInConsole)
            {
                await _moduleDisposer.DisposeAsync(module);
            }
        }
```
`AnsiConsole.Profile.Capabilities.Interactive` evaluates to true, and setting it to false also fixes it, from the same line of code I imagine.

**Comment by thomhurst** [comment]
I thought that might be it. So yeah in interactive mode logs will be output at the end, because otherwise they'd mess up the printing of the progress bars.

However, the fact you're not seeing any progress bars is weird, considering the interactive Boolean is evaluating to true.

This might be an issue with Spectre.Console maybe. But the workaround for now is to set the ShowProgressInConsole Boolean to false.

I'm also going to simplify those checks so it just looks at the ShowProgressInConsole bool.

**Comment by MattParkerDev** [comment]
Should one see progress bars if they don't have an estimated time provider registered?

I might keep digging as to why the progress bars don't show up :)

**Comment by thomhurst** [comment]
Yeah, the first time you run them they will just predict a dummy value of 2 minutes each. If you then run again and you're on the same machine with persistent storage, they'll store a little json file of the module name and how long it last took to complete in your appdata folder, and then use that on subsequent runs to give you a more accurate time.

So I've no idea why they're not showing unfortunately!

Is your terminal window large enough? I don't know if Spectre might not render things if it doesn't deem there to be enough space?

**Comment by MattParkerDev** [comment]
Here's a minimum repro :)

https://github.com/MattParkerDev/Repro.ModularPipelines.InteractiveConsole

**Comment by thomhurst** [comment]
@MattParkerDev  thanks! It actually was a ModularPipelines bug that I'd introduced when refactoring.

I've issued a fix for the latest version. Hopefully you'll see progress bars now.

**Comment by MattParkerDev** [comment]
Fixed, thank you! (2.30.5)
